src/python/pants/core/util_rules/environments.py

```diff
@@ -363,5 +314,106 @@ def extract_process_config_from_environment(tgt: EnvironmentTarget) -> ProcessCo
     return ProcessConfigFromEnvironment(docker_image=docker_image)
 
 
+class EnvironmentSensitiveOptionFieldMixin:
+    subsystem: ClassVar[type[Subsystem]]
+    option_name: ClassVar[str]
+
+
+# Maps between non-list option value types and corresponding fields
+_SIMPLE_OPTIONS: dict[type, type[Field]] = {
+    str: StringField,
+}
+
+# Maps between the member types for list options. Each element is the
+# field type, and the `value` type for the field.
+_LIST_OPTIONS: dict[type, type[Field]] = {
+    str: StringSequenceField,
+}
+
+
+@memoized
+def add_option_fields_for(subsystem: type[Subsystem]) -> Iterable[UnionRule]:
+    """Register environment fields for the `environment_sensitive` options of `subsystem`
+
+    This should be called in the `rules()` method in the file where `subsystem` is defined. It will
+    register the relevant fields under the `local_environment` and `docker_environment` targets.
+    """
+    field_rules: set[UnionRule] = set()
+
+    for option_attrname, option in _collect_options_info_extended(subsystem):
+        if option.flag_options["environment_sensitive"]:
+            field_rules.update(_add_option_field_for(subsystem, option, option_attrname))
+
+    return field_rules
+
+
+def _add_option_field_for(
+    subsystem_t: type[Subsystem], option: OptionsInfo, attrname: str
+) -> Iterable[UnionRule]:
+    option_type: type = option.flag_options["type"]
+    scope = subsystem_t.options_scope
+
+    # If you get a `KeyError` due to a missing value in `map` or `list_map`, it means there
+    # isn't a mapping for a given option type added in the dicts defined earlier in this file.
+    # Add it there (note that currently it's not immediately clear how to support `Enum` options)
+
+    if option_type != list:
+        field_type = _SIMPLE_OPTIONS[option_type]
+    else:
+        member_type = option.flag_options["member_type"]
+        field_type = _LIST_OPTIONS[member_type]
+
+    # The below class will never be used for static type checking outside of this function.
+    # so it's reasonably safe to use `ignore[name-defined]`. Ensure that all this remains valid
+    # if `_OptionFieldMap` is ever modified.
+    class OptionField(field_type, EnvironmentSensitiveOptionFieldMixin):  # type: ignore[valid-type, misc]
+        # TODO: use envvar-like normalization logic here
+        alias = f"{scope}_{option.flag_names[0][2:]}".replace("-", "_")
+        required = False
+        value: Any
+        help = option.flag_options["help"] or ""
```

Comment:
I think I'm in favor of the `help` instead being something that directly calls out the related subsystem:

> Overrides the default value from the option `[python-boostrap].search_path` when this environment target is used.

Reply:
Agreed.